File: union-find/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        rows, cols = len(grid), len(grid[0])
        visited = set()
        island=0

        def bfs(r,c):
            q = deque()
            q.append((r,c))
            visited.add((r,c))

            while q:
                r,c = q.popleft()
                directions = [[1,0],[-1,0], [0,1],[0,-1]]
                for dr, dc in directions:
                    r,c = r+dr, c+ dc

                    if (r in range(rows) and c in range(cols)) and grid[r][c] == "1" and (r,c) not in visited:
                        q.append((r,c))
                        visited.add((r,c))




        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r,c) not in visited:
                    bfs(r,c)
                    island+=1
        return island 

File: matrix/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0; // If the grid is empty, return 0
        }

        int numIslands = 0; // Initialize the number of islands to 0
        int rows = grid.length;
        int cols = grid[0].length;

        // Iterate through each cell in the grid
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    numIslands++; // Increment the island count
                    dfs(grid, i, j); // Perform DFS to mark all connected '1's as visited
                }
            }
        }

        return numIslands; // Return the total number of islands
    }

    // Helper function to perform DFS
    private void dfs(char[][] grid, int i, int j) {
        int rows = grid.length;
        int cols = grid[0].length;

        // Base case: if the cell is out of bounds or not '1', return
        if (i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] != '1') {
            return;
        }

        grid[i][j] = '0'; // Mark the cell as visited by setting it to '0'

        // Recursively visit all adjacent cells (up, down, left, right)
        dfs(grid, i - 1, j); // Up
        dfs(grid, i + 1, j); // Down
        dfs(grid, i, j - 1); // Left
        dfs(grid, i, j + 1); // Right
    }
}